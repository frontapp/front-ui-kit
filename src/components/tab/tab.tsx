import {ellipsis} from 'polished';
import React, {FC, MouseEventHandler} from 'react';
import styled, {css} from 'styled-components';

import {alphas, greys} from '../../helpers/colorHelpers';
import {fonts, fontWeights} from '../../helpers/fontHelpers';

/*
 * Props.
 */

interface TabProps {
  /** The name of the tab. */
  name: string;
  /** Whether the tab is selected. */
  isSelected?: boolean;
  /** Called when the tab is clicked. */
  onClick?: MouseEventHandler;
}

/*
 * Style.
 */

interface StyledTabWrapperDivProps {
  $isSelected?: boolean;
}

const StyledTabWrapperDiv = styled.div<StyledTabWrapperDivProps>`
  display: inline-block;
  flex: 1;
  font-family: ${fonts.system};
  position: relative;
  font-weight: ${fontWeights.semibold};
  color: ${greys.shade60};
  overflow: hidden;
  cursor: default;
  text-align: center;

  ${p => p.$isSelected && css`
    color: ${greys.black};
  `}

  &:hover {
    color: ${greys.black};
  }
`;

const StyledTabNameDiv = styled.div`
  margin-bottom: 12px;
  ${ellipsis()}; // TODO: Remove this when we have tooltip support.
`;

interface StyledSelectedBorderDivProps {
  $isSelected?: boolean;
}

const StyledSelectedBorderDiv = styled.div<StyledSelectedBorderDivProps>`
  background: ${alphas.transparent};
  height: 4px;
  border-top-left-radius: 100px;
  border-top-right-radius: 100px;

  ${p => p.$isSelected && css`
    background: ${greys.black};
  `}
`;

/*
 * Component.
 */

// TODO: Add tooltip support for overflowing names.
export const Tab: FC<TabProps> = props => {
  const {name, isSelected, onClick} = props;
  return (
    <StyledTabWrapperDiv $isSelected={isSelected} onClick={onClick}>
      <StyledTabNameDiv>{name}</StyledTabNameDiv>
      <StyledSelectedBorderDiv $isSelected={isSelected} />
    </StyledTabWrapperDiv>
  );
};
