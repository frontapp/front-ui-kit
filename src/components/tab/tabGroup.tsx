import React, {FC} from 'react';
import styled from 'styled-components';

import {renderChildrenSpecifiedComponents} from '../../helpers/renderHelpers';

/*
 * Props.
 */

interface TabGroupProps {
  /** Tabs to render. Only will render the Tab component. */
  children?: React.ReactNode;
  /** Max width of the group. Defaults to 100% of parent. */
  maxWidth?: number;
}

/*
 * Style.
 */

interface StyledTabGroupWrapperDivProps {
  $maxWidth?: number;
}

const StyledTabGroupWrapperDiv = styled.div<StyledTabGroupWrapperDivProps>`
  display: flex;
  flex-flow: row;
  gap: 16px;
  max-width: ${(p) => (p.$maxWidth ? `${p.$maxWidth}px` : '100%')};
`;

/*
 * Component.
 */

export const TabGroup: FC<TabGroupProps> = ({children, maxWidth}) => {
  const tabGroupContent = renderChildrenSpecifiedComponents(children, ['Tab']);
  return <StyledTabGroupWrapperDiv $maxWidth={maxWidth}>{tabGroupContent}</StyledTabGroupWrapperDiv>;
};
