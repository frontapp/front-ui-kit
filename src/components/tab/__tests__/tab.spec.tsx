import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import {Tab} from '../tab';
import {TabGroup} from '../tabGroup';

describe('<Button /> Spec', () => {
  describe('Snapshots', () => {
    it('should render tab', () => {
      const view = render(<Tab name="Example Tab" />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render selected tab', () => {
      const view = render(<Tab name="Example Tab" isSelected />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render tab group', () => {
      const view = render(
        <TabGroup>
          <Tab name="Example Tab 1" isSelected />
          <Tab name="Example Tab 2" />
          <Tab name="Example Tab 3" />
        </TabGroup>
      );
      expect(view.container).toMatchSnapshot();
    });
  });

  it('should issue onClick events when clicked', async () => {
    const user = userEvent.setup();
    const onClickMock = jest.fn();
    const view = render(<Tab name="Example Tab" onClick={onClickMock} />);

    await user.click(view.getByText('Example Tab'));

    expect(onClickMock).toBeCalled();
  });
});
