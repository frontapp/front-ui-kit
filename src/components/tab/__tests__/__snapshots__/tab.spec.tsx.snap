// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Spec Snapshots should render selected tab 1`] = `
.c0 {
  display: inline-block;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  position: relative;
  font-weight: 600;
  color: #9b9c9e;
  overflow: hidden;
  cursor: default;
  text-align: center;
  color: #001B38;
}

.c0:hover {
  color: #001B38;
}

.c1 {
  margin-bottom: 12px;
  display: inline-block;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
}

.c2 {
  background: transparent;
  height: 4px;
  border-top-left-radius: 100px;
  border-top-right-radius: 100px;
  background: #001B38;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Example Tab
    </div>
    <div
      class="c2"
    />
  </div>
</div>
`;

exports[`<Button /> Spec Snapshots should render tab 1`] = `
.c0 {
  display: inline-block;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  position: relative;
  font-weight: 600;
  color: #9b9c9e;
  overflow: hidden;
  cursor: default;
  text-align: center;
}

.c0:hover {
  color: #001B38;
}

.c1 {
  margin-bottom: 12px;
  display: inline-block;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
}

.c2 {
  background: transparent;
  height: 4px;
  border-top-left-radius: 100px;
  border-top-right-radius: 100px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Example Tab
    </div>
    <div
      class="c2"
    />
  </div>
</div>
`;

exports[`<Button /> Spec Snapshots should render tab group 1`] = `
.c1 {
  display: inline-block;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  position: relative;
  font-weight: 600;
  color: #9b9c9e;
  overflow: hidden;
  cursor: default;
  text-align: center;
  color: #001B38;
}

.c1:hover {
  color: #001B38;
}

.c4 {
  display: inline-block;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  position: relative;
  font-weight: 600;
  color: #9b9c9e;
  overflow: hidden;
  cursor: default;
  text-align: center;
}

.c4:hover {
  color: #001B38;
}

.c2 {
  margin-bottom: 12px;
  display: inline-block;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
}

.c3 {
  background: transparent;
  height: 4px;
  border-top-left-radius: 100px;
  border-top-right-radius: 100px;
  background: #001B38;
}

.c5 {
  background: transparent;
  height: 4px;
  border-top-left-radius: 100px;
  border-top-right-radius: 100px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  gap: 16px;
  max-width: 100%;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Example Tab 1
      </div>
      <div
        class="c3"
      />
    </div>
    <div
      class="c4"
    >
      <div
        class="c2"
      >
        Example Tab 2
      </div>
      <div
        class="c5"
      />
    </div>
    <div
      class="c4"
    >
      <div
        class="c2"
      >
        Example Tab 3
      </div>
      <div
        class="c5"
      />
    </div>
  </div>
</div>
`;
